### PR TITLE
LSPS0 Unified Version Negotiation

### PR DESCRIPTION
**Pull request by SeverinAlexB**
As agreed in the latest meeting (13th of September), this PR standardizes and unifies the version negotiation for all LSPS.

### Why
Every LSPS includes it's own versioning mechanism anyway so we might as well standardize it. A unified mechanism reduces error sources and complexity. It removes the need for version handling in the LSPS itself. It reduces the number of method calls from one for each LSPS to a single call.

### PR Content

- Adds the method `lsps0.select_versions` to select versions for all LSPS.
- Updates `lsps0.list_protocols` to make it similar to `lsps0.select_versions`.
- Adds a version number to LSPS0 itself. This adds a clear upgrade mechanism in case we need to change anything here. Think of max MTU increase, compression/TLV, or even a complete change in protocol.
- Removes all versioning from the existing LSPS.

### CODE REVIEW

**Comment by ErikDeSmedt** [comment]
I support the idea of having a unified method of handling versions across all LSPS.

However, the `lsps0.select_versions` introduces some kind of connection state.
A key-feature of JSON-RPC 2.0 is statelessness and it feels wrong to throw it away on the first call.

The state introduces some extra complexity here. The client and server must store the version-state somewhere and implement hooks to forget the state on connect/disconnect.

It complicates debugging in some way. Connections on my node are usually long-lived. When debugging I don't want to have to browser through old logs to see on what version I am.

To my understanding switching versions would require nodes to fully disconnect. This might affect gossip, payment routing, etc...

I would prefer a solution in which the client asks the server what versions they support. The client could include the `api_version` in every call.

**Comment by SeverinAlexB** [comment]
> The state introduces some extra complexity here. The client and server must store the version-state somewhere and implement hooks to forget the state on connect/disconnect.

I assume you are looking at this from a CLN point of view to develop each LSPS as an individual plugin, correct?


> To my understanding switching versions would require nodes to fully disconnect. This might affect gossip, payment routing, etc...

Good point

**Comment by 2Fast2BCn** [comment]
I argued the exact same thing on Telegram the other day.
It's unclear to me why the LSP needs to know the `version` before the first request comes in.

**Comment by 2Fast2BCn** [comment]
Since we are completely refactoring versioning anyway I'm wondering if we also need to discuss feature bits. I haven't looked into anything other then LSPS1. Feature bits are not needed for that one.

**Comment by SeverinAlexB** [comment]
> If both apps have a version-conflict the LSP-server can only see a single misbehaving node.

I do not understand this issue

**Comment by ErikDeSmedt** [comment]
I might have worded that poorly.
The core issue is that the LSP-client might not even run on the lightning-node itself.

Suppose I would be using 2 wallet apps and connect them to my own node.
Wallet A might implement LSPS2 v1 and the Wallet B implements LSPS2 v2.

Wallet A might have negotiated a channel with an LSP and keep the connection open.
A few days later Wallet B tries to acquire a channel with the same LSP.
The negotiation fails and Wallet B closes the connection and negotiates the usage of LSPS2 v2.

This where things get dangerous. Wallet A might falsely believe he has negotiated the version correctly and might speak LSPS2 v1 to a server that expect LSPS v2. 

The tricky part here is that from a protocol view the node represents a single client. 
In practice, multiple LSP-client apps might run on top of a single node.

**Comment by 2Fast2BCn** [comment]
What about:

```JSON
{
  "protocols": {
    "lsps0": [
      {
        "number": 1,
        "scope": "request",
       }
     ]
    "lsps1": [
      {
        "number": 1,
        "scope": "request",
       },
      {
        "number": 2,
        "scope": "request",
       }
     ]
    "lsps2": [
      {
        "number": 2,
        "scope": "request",
       },
      {
        "number": 3,
        "scope": "session",
       }
     ]
  }
}
```

only when the scope is `session` you need a secondary call `lsps0.start_session`

You could potentially add a concept of a `session_id` where the client provides a `session_id` so there is the option to mix different sessions with different versions ?
In that case when scope is request then version number is always added to every request
In case of scope session you provide session_id in every request

I can be way off here since I've only been reading and implementing LSPS1 ;-)

**Comment by SeverinAlexB** [comment]
- Added ability to switch versions without closing the peer connection.
- Added section about versioning switching/migration behaviour.

File: LSPS0/README.md [hunk 1]
```diff
@@ -512,26 +514,127 @@ the LSP supports:
...
+Except for `lsps0.select_versions` and `lsps0.list_protocols`, any other LSPS specification method 
+call that does not have a negotiated version number, MUST fail with
+
+* `unsupported_version` (1) - The client and the LSP have not yet
+  negotiated a API version via `lsps0.select_versions`.
+
+`lsps0.select_versions` and `lsps0.list_protocols` are exempt from this error
+as they are needed to negotiate the API versions.
+
+`lsps0.select_versions` MAY be called multiple times during the same connection session.
```
[the remark this answers is not in the document]
Comment (SeverinAlexB) [reply LSPS0/README.md:617]: > As I have not heard a convincing argument yet why we should entirely overhaul the basic assumptions of the protocol (namely one node, one LSP, communicating via JSON-RPC over a single BOLT8/TCP connection), it's a NACK so far from me.

This is not done to support multiple apps. The assumption is still one node communicating with the LSPSpec and one version to be active at a time. The only thing I changed here is that a version upgrade/downgrade can happen without disconnecting/connecting the tcp connection as this could have detrimental effects on other services outside of the LSPSpec.

> namely one node, one LSP, communicating via JSON-RPC over a single BOLT8/TCP connection

Happy to write this base assumption in this PR if you wish. This PR is exclusively reserved for one node. If we were to support multiple apps/signer at a time, @ErikDeSmedt or I would create a new PR. I am not convinced yet this is needed tho and there are still discussions going on. Highly likely we will just stick with one node/app.

[the remark this answers is not in the document]
Comment (SeverinAlexB) [reply LSPS0/README.md:617]: I agree, we need to talk about this. That is why I added the migration section to this PR. I quote:

>Every LSPS SHOULD address version switching/migrations behaviour 
> if a version number is increased or decreased.

> > **Rationale** In case the client switches to a different version,
> > existing states might need to be abandoned or migrated.
> > It is important to clearly define these behaviours to avoid
> > confusion.

The thing is: This migration behaviour would have occurred anyway as soon as we have multiple versions of an LSPS. It doesn't matter if we now switch versions during the tpc-session or when disconnecting/reconnecting.

I personally think this should be addressed spec to spec because different specs have different requirements. Happy to discuss on the next call.

> Btw., on this note we should probably generally start talking about protocol-level compatibility guarantees, i.e., if we think some protocols should be resume operation after reconnect (and what implication that has on protocol-level changes) or if we generally require/assume every 'session' to be entirely independent, so we never have to care about upgrade paths between versions.

So far I assumed the protocol resumes operation after each session/reconnect. I don't see a reason why to do short lived session.

**Comment by ErikDeSmedt** [comment]
I'll use this comment as a draft for an alternative proposal.  I'm not fully familiar with LSPS4 yet so I might be missing some requirements or include inaccuracies. Please point them out so we can improve the proposal.

In WEB-API's it common to include a version indicator on every request. Usually, the path of an API endpoint starts with `/api/v1` to indicate the version number. Once `/api/v2` is released, clients can freely pick what version they use. There is no requirement that a client uses the same version consistently. I believe this pattern has been proven to be effective.

## What changes in lsps0?

We implement the proposed changes `lsps0.list_protocols`. The method will return supported version of each protocol.

We will not implement a `lsps0.select_version` call

## What changes in lsps1?

The client includes the version number on every request. 
We will only need to modify `lsps1.get_info`.

## What changes in lsps2 (aka lsps2 v1) and lsps4 (aka lsps2 v2)?

We will include the version number on every request.

Lsps v1 and lsps2 v2 would be two separate API's that don't interfere with each other (except for maybe some clean-up calls)?

If a client wants to upgrade from LSPS2 v1 to LSPS2 v2 they just start using LSPS2 v2. 
- The client can reserve an lsps2 v2 `scid` and start using it
- All lsps2 v1 `jit_channel_scid`'s will be treated as such. 
    - The negotiated fee-rates will apply
    - The server will not send any `lsps4.you_have_new_ppp`-notifications
    - ...

After some time all `jit_channel_scid`'s will either be expired or used.
- The client can forget about `jit_channel_scid`'s if they are expired
- If `jit_channel_scid` is used it becomes just a normal lightning channel with an alias. The client doesn't need to remember that it is a lsps2 v1 channel.

Once the client has forgotten about all LSPS2 v1 channels it has been successfully updated.

# Why I like this proposal better?

- **lsps0 doesn't force specific semantics**: We do not know yet which requirements future upgrades might have. For some future LSPS the `select_version`-call might be the only appropriate mechanism. For other a mechanism described above might be suited better. We can leave this up to the LSPS.

- **No unexpected version-changes**: After network outage a client might be required to call `select_versions` again. This might result in unexpected up- or downgrades in the middle of the night. Sometimes upgrades go bad and require attention. The operational burden of maintaining an LSP-client should be minimized. Giving clients the opportunity to plan upgrades is an improvement. Some upgrades might require user-approval such as the acceptance of new fee-rates. The current proposal wouldn't allow this.

- **No need to define up- or downgrade paths**: A downgrade from LSPS4 (LSPS2 v2) to LSPS2 v1 looks very hard and probably requires user input. The user might want to accept the LSPS2 v2 fee-rates. This might happen unexpectedly.

- **Allows separate implementations for different parts of the spec** : I understand having a single node that powers multiple apps with multiple signers is controversial. However, that is not the point of this comment. In core-lightning we might want to implement an lsps1-plugin and an lsps2-plugin. They can both negotiate different parts of the spec.

- **Allow upgrading without altering existing agreements** : Each lsps represents an agreement between the client and server. Changing the version alters the semantics of the agreement and might result in new rates. In some cases the client might have negotiated a very good rate. This rate might not be available anymore because of chasing market-conditions. There is a risk that clients refuse to upgrade to keep their agreement. To my understanding this risk isn't present in the current spec. But I think it is likely that future enhancements or versions might suffer from it.

**Comment by JssDWt** [comment]
This proposal handles request/response messages, and fixes the versioning issue for us there.

I have a proposal for notifications too:

For notifications, a client always registers for a "notification session". The client does this by sending a request to the LSP: `lspsx.subscribe_x`, containing the version it wants to use. The LSP will respond with an object containing a session identifier. For the duration of the session, the LSP will broadcast notifications to the client, using the version the client specified in the request. 

A session ends when the client requests to stop notifications for that session id, or when the peer connection is dropped. If there are multiple sessions at the same time, the LSP will broadcast notifications to all currently active sessions. 

This will solve the situation where there's multiple software packages handling client-LSP related things, which may be using different versions of an LSPSx protocol.

**Comment by ZmnSCPxj-jr** [comment]
>multiple software packages handling client-LSP related things

This is the bit I would like to ***strongly*** object to.

As I understand it, the intent of greenlight is:

* Ensure that the greenlight node does not have direct access to private keys, thus requiring that apps handle private keys.
  * As node private key is needed to even establish a connection, this implies that if no individual apps are currently running, the greenlight node MUST drop all connections as well, as it can do nothing anyway if the connections are dropped "naturally" by normal network-being-unreliable reality.  It is better to consistently say "once all apps have disconnected from the greenlight node the greenkight node disconnects all LN connections" than to say "99.9% of the time the greenlight node remains connected but in the 0.1% chance it drops accidentally we cannot make or accept new connections because mmmmmmmm internal detail about how apps hold the private key".
  * If the node private key (used for LN BOLT8 comms) is held by greenlight and only channel signing keys are held by apps... that kinda breaks the assumption inherent in the BOLT 8 spec that the node keypair is ultimately held by the same entity that has access to channel signing keys.  Yes, the greenlight node cannot move funds in the channels, but it can still talk on behalf of the node and do things like spam LSPs with requests that it ultimately cannot handle, which makes me suspicious of greenlight.  Yes, you might open-source the greenlight code and show that the open-source greenlight does not do that --- but that does not mean you guys run the open source version.  There is a reason why "Do not trust, verify" is a thing, and why we do not just do something like "oh is the other side running LND? okay we trust that and never verify their messages because they are running LND which is 100% open-source and we can save CPU cycles by having us human developers validate the LND code instead of having our program validate every message".
* The intent is that the greenlight node liquidity is shared among apps.
  * LSPS protocols are all about *managing liquidity* of the client with the LSP. If you intend to share liquidity among apps in a greenlight node, then you should really also *manage* that liquidity across all apps, not have each app manage liquidity.
  * Having the apps handle communication with the LSP and manage the liquidity directly seems broken to me.  It is like: "okay now we run each plugin in the same memory space because we want to share memory resources among them, however each plugin needs to write its own `malloc`, here is the interface to `sbrk`, have fun".  If you want to share resources among various components, you need a component that actually manages those centrally, not expose LSPS1 directly to your apps.
  * As to exposing LSPS directly to apps, which do you think an app developer would prefer?
    * Oh you have to talk to LSP directly, here is an interface to check if the greenlight node has enough inbound liquidity, here is an interface to the LSPS endpoint, have fun.
    * Oh you want inbound liquidity? Here is an interface where you say how much inbound you are expecting, and the greenlight node will automatically get liquidity from somewhere magical so you only need to issue an invoice with these things we will give you but you fill in the hash+secret+metadata.

In short: architecture issue, NACK.

**Comment by SeverinAlexB** [comment]
@JssDWt Session management on the LSP side brings a TON of complexity that I seriously want to avoid. NACK from me.

@ErikDeSmedt An easy way to make your proposal work: Add the version to the method/notification name. Example: `lsps1.v1.get_info` or `lsps4.v1.you_have_new_ppp`. This basically makes each version independent of each other. There would be no difference between having independent specs like `lsps2` and `lsps4` or independent versions `lsps2.v1` and `lsps2.v2`. Versions and specs can be used independently and simultaneously.
I am still neutral on what we should do but this would be at least an easy alternative.

**Comment by JssDWt** [comment]
> In short: architecture issue, NACK.

You haven't posed anything in this rant that discusses version management. The same problem occurs if there are 2 software packages running on top of a lightning node that use the node's api to communicate. It has nothing to do with key management. People run software on top of their nodes where protocols like the LSP spec are implemented. It seems logical to me that some of these packages use other versions than others.


> Session management on the LSP side brings a TON of complexity that I seriously want to avoid. NACK from me.

I don't see how session management can make things more complex. You're already managing this information as an LSP if you want to do a version session per peer connection. It makes sense for a client to request a specific notification stream for a specific version upon connection.

**Comment by roeierez** [comment]
I agree with @JssDWt. 
LSP client implementations are implemented on top of the lightning node which is only used as a transport layer. Allowing it to be used by multiple lsp clients sounds to me like something we want to achieve, it is impossible to force all of them to use the same protocol version.
This proposal assumes there is only one lsp client per node which is a big limitation in the spec IMO.
Allowing registering for notifications with a registration/session id is something very simple to implement, resulting in a lot cleaner spec with unnecessary limitations.

**Comment by kingonly** [comment]
@ZmnSCPxj-jr I don't understand your response tbh. This isn't a race condition. Greenlight is just an example on why this proposed architecture doesn't hold water. You can have the same node used multiple clients with different versions.

**Comment by SeverinAlexB** [comment]
> I don't see how session management can make things more complex. You're already managing this information as an LSP if you want to do a version session per peer connection. It makes sense for a client to request a specific notification stream for a specific version upon connection.

@JssDWt Let me know if I misunderstand you but with my `lsps.version` proposal, you have ALL supported versions running at the same time. Therefore you don't need any sessions or so. The LSP can send a notification whenever it wants and it's the responsibility of the node/app to process it.

Adding notification subscriptions to the LSPSpec does make the spec more complex?

**Comment by roeierez** [comment]
> @JssDWt Let me know if I misunderstand you but with my `lsps.version` proposal, you have ALL supported versions running at the same time. Therefore you don't need any sessions or so. The LSP can send a notification whenever it wants and it's the responsibility of the node/app to process it.

@SeverinAlexB As far as I understand @JssDWt proposed sessions only to bind a specific session to a version, so notifications can be sent to multiple clients that support different versions simultaneously. 
If you mean that the LSP would send notifications for all its supported versions then I agree that there is no need for sessions as every client should filter its relevant notifications. 
In fact this combined with sending the version on a request basis keeps the lsp protocol stateless without the need for reestablishing connection.

In mobile there are some issues when implementing this spec as:
1. If the node is running on the device,  the node frequently loose connection to its peer resulting in the need to re-establish the version negotiation almost on a request basis.
2. If the node is controlled by the mobile app remotely the app, when goes to foreground, never knows if the peer has been disconnected since last time the version was established which resulting again in version negotiation to ensure the version is set.

So my question is if anyway we need to negotiate the version on a request basis, why not just sending it as part of the request, saving this stateful negotiation phase.

**Comment by JssDWt** [comment]
> There is no versioning negotiation in my `lsps.version` proposal. All versions are always running and can be used.

Sounds good. I think I missed this proposal. Do you have a link?

**Comment by ErikDeSmedt** [comment]
> > There is no versioning negotiation in my `lsps.version` proposal. All versions are always running and can be used.
> 
> Sounds good. I think I missed this proposal. Do you have a link?

It's higher in the comments.
https://github.com/BitcoinAndLightningLayerSpecs/lsp/pull/56#issuecomment-1742864915

**Comment by JssDWt** [comment]
> @ErikDeSmedt An easy way to make your proposal work: Add the version to the method/notification name. Example: `lsps1.v1.get_info` or `lsps4.v1.you_have_new_ppp`. This basically makes each version independent of each other. There is no difference between having independent specs like `lsps2` and `lsps4` or independent versions `lsps2.v1` and `lsps2.v2`. Versions and specs can be used independently and simultaneously.

So we would have a spec that introduces backward compatible changes by default. When changes are made to a spec, they are made backward compatible with previous implementations of the spec. If backward incompatible changes are needed, a new spec (e.g. `lsps2.v2`) is created. Needing only the `list_protocols` call, and no version negotiation, because the versions are basically embedded in the jsonrpc `method`. 

> Let me know if I misunderstand you but with my lsps.version proposal, you have ALL supported versions running at the same time. Therefore you don't need any sessions or so. The LSP can send a notification whenever it wants and it's the responsibility of the node/app to process it.

I think this indeed covers all versioning issues, and there would be no need for 'version sessions'. I understand that comment now.
